Race in distributor/protocols/tdp

Found in #198 by turning on the race detector.

```
$ go test -race github.com/coreos/torus/distributor/protocols/tdp
==================
WARNING: DATA RACE
Write at 0x00c42021a000 by goroutine 55:
  internal/race.WriteRange()
      /usr/local/go/src/internal/race/race.go:49 +0x42
  syscall.Read()
      /usr/local/go/src/syscall/syscall_unix.go:164 +0x9a
  net.(*netFD).Read()
      /usr/local/go/src/net/fd_unix.go:216 +0x146
  net.(*conn).Read()
      /usr/local/go/src/net/net.go:173 +0x96
  net.(*TCPConn).Read()
      <autogenerated>:69 +0x72
  github.com/coreos/torus/distributor/protocols/tdp.readConnIntoBuffer()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:130 +0x95
  github.com/coreos/torus/distributor/protocols/tdp.(*Server).handlePutBlock()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:175 +0x526
  github.com/coreos/torus/distributor/protocols/tdp.(*Server).handle()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:108 +0x27a

Previous read at 0x00c42021a000 by goroutine 50:
  internal/race.ReadRange()
      /usr/local/go/src/internal/race/race.go:45 +0x42
  syscall.Write()
      /usr/local/go/src/syscall/syscall_unix.go:182 +0xaa
  net.(*netFD).Write()
      /usr/local/go/src/net/fd_unix.go:298 +0x187
  net.(*conn).Write()
      /usr/local/go/src/net/net.go:185 +0x96
  net.(*TCPConn).Write()
      <autogenerated>:70 +0x72
  github.com/coreos/torus/distributor/protocols/tdp.(*Conn).PutBlock()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/client.go:124 +0x9a3
  github.com/coreos/torus/distributor/protocols/tdp.TestPutBlock()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp_test.go:185 +0x27a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:610 +0xc9

Goroutine 55 (running) created at:
  github.com/coreos/torus/distributor/protocols/tdp.(*Server).serve()
      /home/eric/src/github.com/coreos/torus/distributor/protocols/tdp/tdp.go:80 +0x1c6

Goroutine 50 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:646 +0x52f
  testing.RunTests.func1()
      /usr/local/go/src/testing/testing.go:793 +0xb9
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:610 +0xc9
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:799 +0x4b5
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:743 +0x12f
  main.main()
      github.com/coreos/torus/distributor/protocols/tdp/_test/_testmain.go:80 +0x1b4
==================
--- FAIL: TestPutBlock (2.05s)
	tdp_test.go:187: read tcp 127.0.0.1:32850->127.0.0.1:40000: i/o timeout
FAIL
FAIL	github.com/coreos/torus/distributor/protocols/tdp	4.745s
```

--- a/distributor/protocols/tdp/tdp.go
+++ b/distributor/protocols/tdp/tdp.go
@@ -4,6 +4,7 @@ import (
 	"errors"
 	"io"
 	"net"
+	"sync"
 	"time"
 
 	"github.com/coreos/pkg/capnslog"
@@ -35,9 +36,11 @@ var (
 type Server struct {
 	handler   Handler
 	lst       net.Listener
-	conns     []net.Conn
-	closed    bool
 	blocksize uint64
+
+	mu     sync.RWMutex // protects fields below
+	closed bool
+	conns  []net.Conn
 }
 
 type Handler interface {
@@ -71,12 +74,16 @@ func (s *Server) serve() {
 	for {
 		conn, err := s.lst.Accept()
 		if err != nil {
-			if !s.closed {
+			if !s.isClosed() {
 				clog.Errorf("error listening: %v", err)
 			}
 			return
 		}
+
+		s.mu.Lock()
 		s.conns = append(s.conns, conn)
+		s.mu.Unlock()
+
 		go s.handle(conn)
 	}
 }
@@ -93,7 +100,7 @@ func (s *Server) handle(conn net.Conn) {
 				conn.Close()
 				return
 			}
-			if !s.closed {
+			if !s.isClosed() {
 				clog.Errorf("Error handling: %v", err)
 				conn.Close()
 			}
@@ -115,7 +122,7 @@ func (s *Server) handle(conn net.Conn) {
 			err = errors.New("unknown message on the data port")
 		}
 		if err != nil {
-			if !s.closed {
+			if !s.isClosed() {
 				clog.Errorf("Error handling putblock: %v", err)
 				conn.Close()
 			}
@@ -210,17 +217,29 @@ func (s *Server) handleRebalanceCheck(conn net.Conn, len int, refbuf []byte) err
 	return nil
 }
 
+func (s *Server) isClosed() bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.closed
+}
+
 func (s *Server) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
 	s.closed = true
 	err := s.lst.Close()
 	if err != nil {
 		return err
 	}
+
 	for _, x := range s.conns {
 		err := x.Close()
 		if err != nil {
 			return err
 		}
 	}
+
 	return nil
 }

--- a/distributor/protocols/tdp/tdp_test.go
+++ b/distributor/protocols/tdp/tdp_test.go
@@ -165,9 +165,11 @@ func TestBlockGRPC(t *testing.T) {
 }
 
 func TestPutBlock(t *testing.T) {
-	test := makeTestData(512 * 1024)
+	stest := makeTestData(512 * 1024)
+	ctest := append([]byte(nil), stest...)
+
 	m := &mockBlockRPC{
-		data: test,
+		data: stest,
 	}
 	s, err := Serve("localhost:40000", m, m.BlockSize())
 	if err != nil {
@@ -182,7 +184,7 @@ func TestPutBlock(t *testing.T) {
 	err = c.PutBlock(context.TODO(), torus.BlockRef{
 		INodeRef: torus.NewINodeRef(1, 2),
 		Index:    3,
-	}, test)
+	}, ctest)
 	if err != nil {
 		t.Fatal(err)
 	}

